pypif_sdk/func/calculate_funcs.py

```diff
@@ -0,0 +1,280 @@
+                total_str = ""
+            new_string = symbol[0] + total_str
+            hydrate_string += new_string
+    return formula_string[:hydrate_pos] + hydrate_string
+
+def _create_compositional_array_(expanded_chemical_formaula_string):
+    """
+    Splits an expanded chemical formula string into an array of dictionaries containing information about each element
+    :param expanded_chemical_formaula_string is a clean (not necessarily emperical, but without any special characters) chemical formula string, as returned by _expand_formula_()
+
+    Returns an array of dictionaries
+    """
+    element_array = re.findall('[A-Z][^A-Z]*', expanded_chemical_formaula_string)
+    split_element_array = []
+    for s in element_array:
+        m = re.match(r"([a-zA-Z]+)([0-9\.]*)", s, re.I)
+        if m:
+            items = m.groups()
+            if items[1] == "":
+                items = (items[0], 1)
+        this_e = {"symbol": items[0], "occurances": float(items[1])}
+        
+        split_element_array.append(this_e)
+    return split_element_array
+
+def _consolidate_elemental_array_(elemental_array):
+    """
+    Accounts for non-empirical chemical formulas by taking in the compositional array generated by _create_compositional_array_() and returning a consolidated array of dictionaries with no repeating elements
+
+    :param elemental_array is an elemental array generated from _create_compositional_array_()
+
+    Returns an array of element dictionaries
+    """
+    condensed_array = []
+    for e in elemental_array:
+        exists = False
+        for k in condensed_array:
+            if k["symbol"] == e["symbol"]:
+                exists = True
+                k["occurances"] += e["occurances"]
+        if not exists:
+            condensed_array.append(e)
+    return condensed_array
+
+def _calculate_n_atoms_(elemental_array):
+    """
+    Calculates the total number of atoms in the system.
+    :param elemental_array is an array of dictionaries as generated by either _create_compositional_array_() or _consolidate_elemental_array_()
+    returns the total number of atoms (can be a partial number in the case of alloys or partial hydrates)
+    """
+    n = 0
+    for a in elemental_array:
+        n += a["occurances"]
+    return n
+
+def _add_ideal_atomic_weights_(elemental_array):
+    """
+    Uses mendeleev.element to find the molar mass of the element in question, and then multiplies that by the occurances of the element.
+    Adds the "weight" property to each of the dictionaries in elemental_array
+    :param elemental_array an array of dictionaries containing information about the elements in the system 
+    Returns the appended elemental_array
+    """
+    for a in elemental_array:
+        # this_atomic_weight = element(a["symbol"]).atomic_weight
+        this_atomic_weight = elements_data[a["symbol"]]["atomic_weight"]
+        a["weight"] = round(a["occurances"] * this_atomic_weight, 2)
```

Comment:
Why round? That makes the weight composition non-equivalent to the original chemical formula. The atomic percents are real-valued, anyways.